backend/src/od.rs

```diff
@@ -13,25 +12,36 @@ pub struct ZoneID(pub usize);
 #[derive(Serialize, Deserialize)]
 pub struct DemandModel {
     pub zones: Vec<Zone>,
+    #[serde(skip)]
+    pub cached_zone_roads: Vec<Vec<RoadID>>,
     // (zone1, zone2, count), with count being the number of trips from zone1 to zone2
     pub desire_lines: Vec<(ZoneID, ZoneID, usize)>,
 }
 
 impl DemandModel {
     /// Turn all of the zones into Mercator. Don't do this when originally building and serializing
     /// them, because that process might not use exactly the same Mercator object.
-    pub fn finish_loading(&mut self, mercator: &Mercator) {
+    ///
+    /// Also, calculate all the roads in each zone.
+    pub fn finish_loading(&mut self, map: &MapModel) {
+        self.cached_zone_roads = vec![vec![]; self.zones.len()];
+
+        let mut prepared_zones = vec![];
         for zone in &mut self.zones {
-            mercator.to_mercator_in_place(&mut zone.geometry);
-            let bbox = zone.geometry.bounding_rect().unwrap();
-            zone.x1 = (bbox.min().x * 100.0) as i64;
-            zone.y1 = (bbox.min().y * 100.0) as i64;
-            zone.x2 = (bbox.max().x * 100.0) as i64;
-            zone.y2 = (bbox.max().y * 100.0) as i64;
+            map.mercator.to_mercator_in_place(&mut zone.geometry);
+            prepared_zones.push(PreparedGeometry::from(&zone.geometry));
+        }
+
+        for road in &map.roads {
+            for (zone_idx, zone) in prepared_zones.iter().enumerate() {
+                if zone.relate(&road.linestring).is_intersects() {
+                    self.cached_zone_roads[zone_idx].push(road.id);
```

Comment:
Previously, to generate requests:

1. generate a random point within the zone.geometry bbox
2. do a `contains` check to make sure it was a valid point (else GOTO step 1)
3. Do a nearest neighbor query to get the road nearest that point

step 2 and 3 are both expensive. Now, we do this computation up front to see which roads are in each zone, and trip requests are made by grabbing random roads. It's not perfect but I think it's defensible, and it's faster, especially for the unsampled (more accurate) mode.

fast_sample: 713ms -> 666ms
all samples: 4459ms -> 2989ms